xfer/model_handler/model_handler.py

```diff
@@ -43,103 +44,485 @@ def __init__(self, module, context_function=mx.context.cpu, num_devices=1, data_
+                                                                    symbol_dict, drop_layer_name,
+                                                                    temp_symbol_dict[consts.NODES], del_node_op_idx)
+            if join_deleted:
+                layers_dropped.append(join_layer_name)
+
+            # Update symbol dictionary attributes
+            symbol_dict[consts.ARG_NODES] = self._get_arg_nodes(symbol_dict[consts.NODES])
+            symbol_dict[consts.HEADS] = self._get_heads(symbol_dict[consts.NODES], self._get_output_layer_names(
+                temp_symbol_dict[consts.NODES], temp_symbol_dict[consts.HEADS]))
+            symbol_dict[consts.NODES] = self._update_inputs(symbol_dict[consts.NODES], temp_symbol_dict[consts.NODES],
+                                                            drop_layer_name, join_deleted, join_idx)
+
+        sym = mx.sym.load_json(json.dumps(symbol_dict))
+
+        logging.info('{} deleted from model bottom'.format(', '.join(layers_dropped)))
+        if drop_layer_names is not None:
+            if len(drop_layer_names) > 0:
+                logging.warning('Did not use all of drop_layer_names: {}'.format(', '.join(drop_layer_names)))
         self.update_sym(sym)
 
-    def add_layer_top(self, layer_factory_list):
+    def add_layer_top(self, layer_list):
         """
         Add layer to output of model.
 
-        :param layer_factory: List of LayerFactory objects to be added to model output.
-        :type layer_factory: list(:class:`LayerFactory`)
+        :param layer_list: List of MxNet symbol layers to be added to model output.
+        :type layer_list: list(:class:`mx.symbol`)
         """
         if '_label' in self.symbol.get_internals().list_outputs()[consts.LABEL_IDX]:
             raise exceptions.ModelError('Cannot add layer above output layer')
-        sym = self.symbol
         added_layer_names = []
-        for lf in layer_factory_list:
-            self._validate_layer_name(lf.attributes[consts.NAME])
-            sym = lf.create_layer(sym)
-            added_layer_names.append(lf.attributes[consts.NAME])
+
+        net_symbol_dict = self._get_symbol_dict(self.symbol)
+        for layer_symbol in layer_list:
+            added_layer_names.append(layer_symbol.name)
+            layer_symbol_dict = self._get_symbol_dict(layer_symbol)
+
+            num_existing_nodes = len(net_symbol_dict[consts.NODES])
+
+            # Update inputs of nodes
+            for node in layer_symbol_dict[consts.NODES]:
+                for ip in node[consts.INPUTS]:
+                    ip[0] += num_existing_nodes - 1
+
+            # Concatentate nodes list
+            net_symbol_dict[consts.NODES] = net_symbol_dict[consts.NODES] + layer_symbol_dict[consts.NODES][1:]
+
+            # Update attributes of network symbol dictionary
+            net_symbol_dict[consts.HEADS] = self._get_heads(net_symbol_dict[consts.NODES], layer_symbol.name)
```

Comment:
Couple of notes here:

1. `layer_symbol.name` is passed as a string. This will cause `_get_heads` to misbehave because it ends up doing a substring check instead of checking if an element is present in a list. (Corrected code suggestion below)
2. When making function calls with arguments whose purpose may not be obvious, it's good to include the argument_name so it becomes much easier to read, without having to go to the function definition.

```suggestion
            net_symbol_dict[consts.HEADS] = self._get_heads(nodes=net_symbol_dict[consts.NODES], output_layer_names=[layer_symbol.name])
```

Reply:
Thanks, I will make this change separately because it goes over the line length limit.

I will include argument names elsewhere too where it is not clear.